File: backend/ballotnav/api/dropoffs.py
""" dropoff api """
from flask import Blueprint, current_app
import structlog
from webargs import fields
from webargs.flaskparser import use_args

from ballotnav.models import Dropoff, dropoff_schema

dropoffs = Blueprint("dropoffs", __name__)
logger = structlog.get_logger()


@dropoffs.route("/dropoffs", methods=["GET"])
@use_args({"state_name": fields.Str(), "county": fields.Str()}, location="querystring")
def get_location(args):
    """ get location """
    res = Dropoff.get(dict(county=args.get("county"), state_short_code=args.get("state_name")))

    if not res:
        logger.msg("No locations found", args=args)
        return {"dropoffs": [], "status": "error"}, 404
    return {"dropoffs": dropoff_schema.dump(res, many=True)}, 200


@dropoffs.route("/dropoff", methods=["POST"])
@use_args(
    {
        "state_name": fields.Str(required=True),
        "state_short_code": fields.Str(required=True),
        "county": fields.Str(required=True),
        "position": fields.Str(),
        "contact_name": fields.Str(),
        "address_1": fields.Str(),
        "address_2": fields.Str(),
        "email": fields.Str(),
        "fax": fields.Str(),
        "phone": fields.Str(),
        "county_website": fields.Str(),
        "source_url": fields.Str(),
    }
)
def create_location(args):
    """ create a dropoff location entry """
    logger.info(f"CREATE a location call")

    res = Dropoff.create(args)

    if not res:
        return {"status": "error"}, 400
    return {"status": "ok", "dropoffs": {**dropoff_schema.dump(res)}}, 201

File: backend/models/Dropoffs.js
/*
 * define dropoffs table
 */

const Dropoffs = function (sequelize, DataTypes) {
  return sequelize.define('Dropoffs', {
    state_name :{
        type: DataTypes.STRING(20),
        allowNull: false
    },
    state_short_code: {
        type: DataTypes.STRING(2),
        allowNull: false
    },
    county: {
        type: DataTypes.STRING(100),
        allowNull: false
    },
    position: {
        type: DataTypes.STRING
    }, 
    contact_name: {
        type: DataTypes.STRING
    },
    address_1: {
        type: DataTypes.STRING
    },
    address_2: {
        type: DataTypes.STRING
    }, 
    email: {
        type: DataTypes.STRING
    },
    fax: {
        type: DataTypes.STRING(40)
    },
    phone: {
        type: DataTypes.STRING(40)
    },
    county_website: {
        type: DataTypes.STRING
    },
    source_url: {
        type: DataTypes.STRING
    },
    latitude: {
        type: DataTypes.INTEGER
    }, 
    longitude: {
        type: DataTypes.INTEGER
    } 
  }); 
}

module.exports = Dropoffs;
